node/libs/roles/src/attester/messages/l1_batch.rs

```diff
@@ -0,0 +1,122 @@
+use zksync_concurrency::time;
+
+use crate::{
+    attester::{self, AggregateSignature},
+    validator::Genesis,
+};
+
+use super::{SignedBatchMsg, Signers};
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Default, PartialOrd)]
+/// A batch number.
+pub struct BatchNumber(pub u64);
+
+impl BatchNumber {
+    /// Increment the batch number.
+    pub fn next_batch_number(&self) -> BatchNumber {
+        BatchNumber(self.0.checked_add(1).unwrap_or(0))
+    }
+}
+
+/// A message to send by attesters to the gossip network.
+/// It contains the attester signature to sign the block batches to be sent to L1.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct L1Batch {
+    /// The number of the batch.
+    pub number: BatchNumber,
+    /// Time at which this message has been signed.
+    pub timestamp: time::Utc,
+}
+
+/// A certificate for a batch of L2 blocks to be sent to L1.
+/// It contains the signatures of the attesters that signed the batch.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct L1BatchQC {
+    /// The aggregate signature of the signed L1 batches.
+    pub signature: AggregateSignature,
+    /// The attesters that signed this message.
+    pub signers: Signers,
+    /// The message that was signed.
+    pub message: L1Batch,
+}
+
+/// Error returned by `L1BatchQC::verify()` if the signature is invalid.
+#[derive(thiserror::Error, Debug)]
+pub enum L1BatchQCVerifyError {
+    /// Bad signature.
+    #[error("bad signature: {0:#}")]
+    BadSignature(#[source] anyhow::Error),
+    /// Not enough signers.
+    #[error("not enough signers: got {got}, want {want}")]
+    NotEnoughSigners {
+        /// Got signers.
+        got: u64,
+        /// Want signers.
+        want: u64,
+    },
+    /// Bad signer set.
+    #[error("signers set doesn't match genesis")]
+    BadSignersSet,
+}
+
+impl L1Batch {
+    /// Checks if `self` is a newer version than `b`.
+    pub fn is_newer(&self, b: &Self) -> bool {
+        (&self.number, self.timestamp) > (&b.number, b.timestamp)
+    }
+}
+
+impl L1BatchQC {
+    /// Create a new empty instance for a given `L1Batch` message.
+    pub fn new(message: L1Batch, genesis: &Genesis) -> Self {
+        Self {
+            message,
+            signers: Signers::new(genesis.attesters_committee.len()),
+            signature: attester::AggregateSignature::default(),
+        }
+    }
+
+    /// Add a attester's signature.
+    /// Signature is assumed to be already verified.
+    pub fn add(&mut self, msg: &SignedBatchMsg<L1Batch>, genesis: &Genesis) {
+        if self.message != msg.msg {
```

Comment:
nit: ptal at CommitQC - `add()` returns an error now, rather than silently ignoring useless signatures.

Reply:
Got it, I made some extra changes to the code, incorporating `anyhow` to utilize some helpful macros. Let me know if you prefer the previous code, or if you prefer to mimic this approach in the validators' code as well.